test/clonevolume_tester.py

```diff
@@ -0,0 +1,262 @@
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+
+
+# TODO: Rollback is needed for created volume else unit test would fail
+class TestCloneDefaultEtcdSaveFails(CloneVolumeUnitTest):
+    pass
+
+
+# Offline copy
+class TestCloneOfflineCopy(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.createVolume.assert_called()
+        mock_3parclient.copyVolume.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME,
+                         # Difference in size of source and cloned volume
+                         # triggers offline copy. Src volume size is 2.
+                         "size": 20}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        mock_3parclient.getTask.return_value = {'status': data.TASK_DONE}
+
+
+# Make copyVolume operation fail
+class TestCloneOfflineCopyFails(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+        # TODO: Check following were invoked
+        # self.client.createVolumeSet(vvs_name, domain)
+        # self._set_flash_cache_policy_in_vvs(flash_cache, vvs_name)
+        # self.client.addVolumeToVolumeSet(vvs_name, volume_name)
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME,
+                         # Difference in size of source and cloned volume
+                         # triggers offline copy. Src size is 2.
+                         "size": 20}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.getCPG.return_value = {}
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        # TASK_FAILED simulates failure of copyVolume() operation
+        mock_3parclient.getTask.return_value = {'status': data.TASK_FAILED}
+
+
+class TestCloneInvalidSourceVolume(CloneVolumeUnitTest):
```

Comment:
Here proper Err response needs to be returned
This as well looks incomplete test case

Reply:
Okay